backend/typescript/services/implementations/reviewService.ts

```diff
@@ -32,6 +37,156 @@ class ReviewService implements IReviewService {
     if (db !== sequelize) sequelize.close(); // Using test db instead of main db
   }
 
+  /* eslint-disable class-methods-use-this, no-await-in-loop */
+  async findOrCreateTag(
+    review: PgReview,
+    tag: TagRequest,
+    t: Transaction,
+  ): Promise<PgTag> {
+    const newTag = await PgTag.findOrCreate({
+      where: { name: tag.name },
+      transaction: t,
+    }).then((data) => data[0]);
+    await review.$add("tags", newTag, { transaction: t });
+    return newTag;
+  }
+
+  async findOrCreateTags(
+    review: PgReview,
+    tags: TagRequest[],
+    t: Transaction,
+  ): Promise<TagResponse[]> {
+    const tagsRet: TagResponse[] = [];
+    for (let i = 0; i < tags.length; i += 1) {
+      const tag = await this.findOrCreateTag(review, tags[i], t);
+      tagsRet.push({ id: tag.id, name: tag.name });
+    }
+    return tagsRet;
+  }
+
+  async findOrCreateSeries(
+    seriesName: string | null | undefined,
+    t: Transaction,
+  ): Promise<PgSeries | null> {
+    let series = null;
+    if (seriesName) {
+      series = await PgSeries.findOrCreate({
+        where: { name: seriesName },
+        transaction: t,
+      }).then((data) => data[0]);
+    }
+
+    return series;
+  }
+
+  async findOrCreateAuthor(
+    book: PgBook,
+    author: AuthorRequest,
+    t: Transaction,
+  ): Promise<PgAuthor> {
+    const newAuthor = await PgAuthor.findOrCreate({
+      where: {
+        full_name: author.fullName,
+        display_name: author.displayName || null,
+        attribution: author.attribution || null,
+      },
+      transaction: t,
+    }).then((data) => data[0]);
+    await book.$add("authors", newAuthor, { transaction: t });
+    return newAuthor;
+  }
+
+  async findOrCreatePublisher(
+    book: PgBook,
+    publisher: PublisherRequest,
+    t: Transaction,
+  ): Promise<PgPublisher> {
+    const pub = await PgPublisher.findOrCreate({
+      where: {
+        full_name: publisher.fullName,
+        publish_year: publisher.publishYear,
+      },
+      transaction: t,
+    }).then((data) => data[0]);
+    await book.$add("publishers", pub, { transaction: t });
+    return pub;
+  }
+
+  async createBook(
+    review: PgReview,
+    book: BookRequest,
+    t: Transaction,
+  ): Promise<BookResponse> {
+    const series = book.series
+      ? await this.findOrCreateSeries(book.series.name, t)
+      : undefined;
+
+    const newBook = await PgBook.create(
+      {
+        review_id: review.id,
+        cover_image: book.coverImage,
+        title_prefix: book.titlePrefix,
+        title: book.title,
+        series_id: series?.id || null,
+        series_order: book.seriesOrder || null,
+        illustrator: book.illustrator || null,
+        translator: book.translator || null,
+        formats: book.formats,
+        age_range: [book.minAge, book.maxAge],
+      },
+      { transaction: t },
+    );
```

Comment:
can we break this block into a helper function?

Reply:
What would the helper function do? We would still need to call a create book function where we would still pass the fields as above so we won't be really abstracting functionality.